pkg/cli/gettoken/jwt/jwt.go

```diff
@@ -0,0 +1,68 @@
+package jwt
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// Claims represents claims of an ID token.
+type Claims struct {
+	Subject string
+	Expiry  time.Time
+	Pretty  string // string representation for debug and logging
+}
+
+// IsExpired returns true if the token is expired.
+func (c *Claims) IsExpired(t time.Time) bool {
+	return c.Expiry.Before(t)
+}
+
+// DecodeWithoutVerify decodes the JWT string and returns the claims.
+// Note that this method does not verify the signature and always trust it.
+func DecodeWithoutVerify(s string) (*Claims, error) {
+	payload, err := DecodePayloadAsRawJSON(s)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode the payload: %w", err)
+	}
+	var claims struct {
+		Subject   string `json:"sub,omitempty"`
+		ExpiresAt int64  `json:"exp,omitempty"`
+	}
+	if err := json.NewDecoder(bytes.NewReader(payload)).Decode(&claims); err != nil {
+		return nil, fmt.Errorf("could not decode the json of token: %w", err)
+	}
+	var prettyJson bytes.Buffer
+	if err := json.Indent(&prettyJson, payload, "", "  "); err != nil {
+		return nil, fmt.Errorf("could not indent the json of token: %w", err)
+	}
+	return &Claims{
+		Subject: claims.Subject,
+		Expiry:  time.Unix(claims.ExpiresAt, 0),
+		Pretty:  prettyJson.String(),
+	}, nil
+}
+
+// DecodePayloadAsRawJSON extracts the payload and returns the raw JSON.
+func DecodePayloadAsRawJSON(s string) ([]byte, error) {
+	parts := strings.SplitN(s, ".", 3)
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("wants %d segments but got %d segments", 3, len(parts))
+	}
+	payloadJSON, err := decodePayload(parts[1])
+	if err != nil {
+		return nil, fmt.Errorf("could not decode the payload: %w", err)
+	}
+	return payloadJSON, nil
+}
+
+func decodePayload(payload string) ([]byte, error) {
+	b, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(payload)
+	if err != nil {
+		return nil, fmt.Errorf("invalid base64: %w", err)
+	}
+	return b, nil
+}
```

Comment:
alternatively explain why we need it (maybe to reduce roundtrips to the provider? Could we cache the keys for verification instead?). Aren't there libraries that actually do these kinds of things?

Reply:
I'm not sure about why author didn't choose to use `oidc.Provider.Verify` instead of implementing https://github.com/int128/kubelogin/blob/b1d8e8f7e169cddd7ec3ad9530095209ca654877/pkg/jwt/decode.go#L3. Is it due to a subtle bug, or just a preference?. I'd not take any risk about changing this until we have good CI coverage and sufficient testing to ensure that when we change this to `oidc.Provider.Verify()`, everything will work.

Reply:
CI coverage won't help if you're worrying about roundtrips to the provider. Please check whether there are more standard libraries that either do this work or that are source of it.
Is there a reason you did not vendor this code from the original?

Reply:
Updated the code to always verify token with `oidc.Provider.Verify` and entirely removed this jwt package.